crates/linker/src/macho.rs

```diff
@@ -601,453 +630,292 @@ fn gen_macho_le(
+    let start_of_roc_commands = {
+        let mut out_offset = 0;
+        let mut in_offset = 0;
+        let mut size = md.linkedit_segment;
+
+        out_mmap[out_offset..out_offset + size]
+            .copy_from_slice(&exec_data[in_offset..in_offset + size]);
+        out_offset += size;
+        in_offset += size;
+
+        // Make space for Roc segment load commands
+        let start_of_roc_commands = out_offset;
+        out_offset += required_size;
+        md.linkedit_segment += required_size;
+
+        // "Delete" the dylib load command - by copying all the bytes before it
+        // and all the bytes after it, while skipping over its bytes.
+        // It has a dynamic-length string at the end that we also need to delete,
+        // in addition to the header.
+        size = macho_load_so_offset - in_offset;
+        out_mmap[out_offset..out_offset + size]
+            .copy_from_slice(&exec_data[in_offset..in_offset + size]);
+        out_offset += size;
+        in_offset += size;
+
+        // Skip "deleted" load command
+        in_offset += total_cmd_size;
+
+        size = end_of_cmds - in_offset;
+        out_mmap[out_offset..out_offset + size]
+            .copy_from_slice(&exec_data[in_offset..in_offset + size]);
+
+        // Copy the rest of the file.
+        out_mmap[md.start_of_first_section..]
+            .copy_from_slice(&exec_data[md.start_of_first_section..]);
+
+        start_of_roc_commands
+    };
 
-    // TODO: this needs to change to adding the 2 new commands when we are ready.
-    // -1 because we're deleting 1 load command and then NOT adding 2 new ones.
+    // Add Roc segment commands with segments size 0.
     {
-        let added_bytes = -(total_cmd_size as isize); // TODO: Change when add the new sections.
-        out_header.ncmds.set(LE, num_load_cmds - 1);
-        out_header
-            .sizeofcmds
-            .set(LE, (size_of_cmds as isize + added_bytes) as u32);
+        let mut offset = start_of_roc_commands;
+
+        md.roc_text_segment = offset;
+        set_segment_load_command(&mut out_mmap, offset, b"__ROC_TEXT", E | R, false, 1);
+        offset += mem::size_of::<macho::SegmentCommand64<LE>>();
+        set_section_load_command(
+            &mut out_mmap,
+            offset,
+            b"__text",
+            b"__ROC_TEXT",
+            macho::S_ATTR_PURE_INSTRUCTIONS | macho::S_ATTR_SOME_INSTRUCTIONS,
+        );
+        offset += mem::size_of::<macho::Section64<LE>>();
+
+        md.roc_data_const_segment = offset;
+        set_segment_load_command(&mut out_mmap, offset, b"__ROC_DATA_CONST", W | R, true, 1);
+        offset += mem::size_of::<macho::SegmentCommand64<LE>>();
+        set_section_load_command(
+            &mut out_mmap,
+            offset,
+            b"__const",
+            b"__ROC_DATA_CONST",
+            macho::S_REGULAR,
+        );
+        offset += mem::size_of::<macho::Section64<LE>>();
+
+        md.roc_data_segment = offset;
```

Comment:
Just curious, have you seen use of this segment?

I don't think we map the data section in elf current, just read only data. That said, I think some of compiler rt requires it, so we probably need to add it.

Reply:
Cannot confirm if this is truly needed, but it looks like you are right that compiler-rt might require it and in particular stack protection mechanisms. Here's a listing of `__DATA,__data` symbols from the `echo.roc` app in the relocatable object file:

```
$ zacho -s tmp_roc.o | grep __DATA,__data
  0x000000000000a230 (__DATA,__data) non-external ltmp1
  0x000000000000a230 (__DATA,__data) weak external ___stack_chk_guard
  0x000000000000a238 (__DATA,__data) weak external __mh_execute_header
```